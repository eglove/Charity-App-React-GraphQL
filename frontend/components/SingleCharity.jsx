import React, {Component} from 'react';
import gql from 'graphql-tag';
import Link from 'next/link';
import {Query} from "react-apollo";
import styled from "styled-components";
import Head from 'next/head';
import Error from './ErrorMessage';

const SingleCharityStyles = styled.div`
    max-width: 1200px;
    margin: 2em auto;
    box-shadow: ${props => props.theme.bs};
    min-height: 800px;
    img {
        object-fit: contain;
        display: block;
        margin-left: auto;
        margin-right: auto;
        margin-top: 5px;
        width: 300px;
    }
    a {
        text-decoration: underline;
    }
    .details {
        margin: 3em;
        font-size: 1em;
    }
`;

const SINGLE_CHARITY_QUERY = gql`
    query SINGLE_CHARITY_QUERY($id: ID!) {
        charity(where: {id: $id}) {
            id
            ein
            name
            website
            description
            largeImage
            street
            city
            state
            zip
        }
    }
`;

class SingleCharity extends Component {
    render() {
        return (
            <Query query={SINGLE_CHARITY_QUERY} variables={{
                id: this.props.id
            }}>
                {({error, loading, data}) => {
                    if (error) return <Error error={error}/>;
                    if (loading) return <p>Loading...</p>;
                    if (!data.charity) return <p>No Charity for ID: {this.props.id}</p>
                    const charity = data.charity;
                    return (
                        <SingleCharityStyles>
                            <Head>
                                <title>Cognitame ðŸ’™ {charity.name}</title>
                            </Head>
                            <img src={charity.largeImage} alt={charity.imageDescription}/>
                            <div className="details">
                                <p>
                                    {charity.name}
                                    &emsp;
                                    <Link href={{
                                        pathname: 'https://apps.irs.gov/app/eos/allSearch.do',
                                        query: {
                                            ein1: charity.ein,
                                            resultsPerPage: '25',
                                            indexOfFirstRow: '0',
                                            dispatchMethod: 'searchAll',
                                        }
                                    }}>
                                        <a>{charity.ein}</a>
                                    </Link>
                                    &emsp;
                                    {charity.website && (
                                        <>
                                            <Link href={charity.website}>
                                                <a>Website</a>
                                            </Link>
                                            &emsp;
                                        </>
                                    )}
                                    <Link href={{
                                        pathname: 'https://smile.amazon.com/gp/chpf/homepage',
                                        query: {q: charity.ein}
                                    }}>
                                        <a>AmazonSmile</a>
                                    </Link>
                                </p>
                                <p>{charity.description}</p>
                                <p>{charity.street}<br/>{charity.city}, {charity.state} {charity.zip}</p>
                            </div>
                        </SingleCharityStyles>
                    )
                }}
            </Query>
        );
    }
}

export default SingleCharity;
export {SINGLE_CHARITY_QUERY};