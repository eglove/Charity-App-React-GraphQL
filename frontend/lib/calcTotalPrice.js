export default function calcTotalPrice(favorites) {
  return favorites.reduce((tally, favorite) => {
    if (!favorite.charity) return tally;
    return tally + favorite.amount;
  }, 0);
}
