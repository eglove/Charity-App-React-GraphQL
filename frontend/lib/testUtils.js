import casual from 'casual';

// seed it so we get consistent results
casual.seed(777);

const fakeCharity = () => ({
  __typename: 'Charity',
  id: casual.uuid,
  ein: casual.uuid,
  name: casual.company_name,
  description: casual.description,
  website: casual.url,
  image: `${casual.word}.png`,
  largeImage: `${casual.word}.png`,
  imageDescription: casual.description,
  street: casual.street,
  city: casual.city,
  state: casual.state,
  zip: casual.zip,
  user: casual.uuid,
});

const fakeUser = () => ({
  __typename: 'User',
  id: casual.uuid,
  name: casual.name,
  email: casual.email,
  password: casual.password,
  totalDonated: casual.random,
  permissions: ['ADMIN'],
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  LocalStorageMock,
  fakeCharity,
  fakeUser,
};
