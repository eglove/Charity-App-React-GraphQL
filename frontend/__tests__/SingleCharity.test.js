import {mount} from "enzyme";
import wait from "waait";
import SingleCharity, {SINGLE_CHARITY_QUERY} from "../components/SingleCharity";
import {MockedProvider} from "react-apollo/test-utils";
import {fakeCharity} from "../lib/testUtils";
import toJson from "enzyme-to-json";

describe('<SingleCharity/>', () => {
    it('renders with proper data', async () => {
        const mocks = [
            {
                // when someone makes request with this query & variable
                request: {query: SINGLE_CHARITY_QUERY, variables: {id: '123'}},
                // return this fake data
                result: {
                    data: {
                        charity: fakeCharity(),
                    },
                },
            },
        ];

        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <SingleCharity id="123"/>
            </MockedProvider>
        );

        expect(wrapper.text()).toContain('Loading...');

        // Wait and update wrapper after query finishes
        await wait();
        wrapper.update();

        expect(toJson(wrapper.find('img'))).toMatchSnapshot();
        expect(toJson(wrapper.find('p'))).toMatchSnapshot();
    });

    it('Errors with a not found charity', async () => {
        const mocks = [
            {
                request: {query: SINGLE_CHARITY_QUERY, variables: {id: '123'}},
                result: {
                        errors: [{message: 'Charity Not Found!'}],
                },
            },
        ];

        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <SingleCharity id="123"/>
            </MockedProvider>
        );

        await wait();
        wrapper.update();
        // console.log(wrapper.debug());
        const charity = wrapper.find('[data-test="graphql-error"]');
        expect(charity.text()).toContain('Charity Not Found!');
        expect(toJson(charity)).toMatchSnapshot();
    });
});