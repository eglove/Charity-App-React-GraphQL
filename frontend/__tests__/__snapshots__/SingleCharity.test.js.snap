// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleCharity/> Errors with a not found charity 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Oh no!
  </strong>
  Charity Not Found!
</p>
`;

exports[`<SingleCharity/> renders with proper data 1`] = `
<img
  src="nemo.png"
/>
`;

exports[`<SingleCharity/> renders with proper data 2`] = `
Array [
  <p>
    Bartoletti Group
     
    <Link
      href={
        Object {
          "pathname": "https://apps.irs.gov/app/eos/allSearch.do",
          "query": Object {
            "dispatchMethod": "searchAll",
            "ein1": "7ae8c415-d3f2-4368-8393-f090492fd03e",
            "indexOfFirstRow": "0",
            "resultsPerPage": "25",
          },
        }
      }
    >
      <a
        href="https://apps.irs.gov/app/eos/allSearch.do?ein1=7ae8c415-d3f2-4368-8393-f090492fd03e&resultsPerPage=25&indexOfFirstRow=0&dispatchMethod=searchAll"
        onClick={[Function]}
      >
        7ae8c415-d3f2-4368-8393-f090492fd03e
      </a>
    </Link>
     
    <Link
      href="http://www.Evie.io/"
    >
      <a
        href="http://www.Evie.io/"
        onClick={[Function]}
      >
        Website
      </a>
    </Link>
     
    <Link
      href={
        Object {
          "pathname": "https://smile.amazon.com/gp/chpf/homepage",
          "query": Object {
            "q": "7ae8c415-d3f2-4368-8393-f090492fd03e",
          },
        }
      }
    >
      <a
        href="https://smile.amazon.com/gp/chpf/homepage?q=7ae8c415-d3f2-4368-8393-f090492fd03e"
        onClick={[Function]}
      >
        AmazonSmile
      </a>
    </Link>
  </p>,
  <p>
    Vel provident dicta perferendis dolorum cupiditate illum. Ut assumenda sint quia sed exercitationem omnis laborum exercitationem. Sapiente laudantium molestias assumenda quasi adipisci mollitia iste. Omnis debitis rem similique temporibus aliquam sed sapiente soluta corporis.
  </p>,
  <p>
    Helene Rest
    <br />
    Rocioville
    , 
    Michigan
     
  </p>,
]
`;
